File: kotlin/src/main/kotlin/com/bhengubv/circleai/companion/reasoning/WorldModel.kt
// WorldModel.kt
//
// Kotlin port of the CircleAI.Companion world-model contract and its two
// concrete implementations — the C# reference is the exact spec:
//   IWorldModel            (HerJarvisContracts.cs, contract 5)
//   CausalPrediction       (record)
//   FrequencyWorldModel    (HerJarvisRealImplementations.cs, impl 5)
//   BayesianWorldModel     (BayesianWorldModel.cs, Phase E3)
//
// A world model learns P(outcome | observations) from registered evidence and
// predicts the most probable outcome for a scenario. The scenario is a JSON
// object; each property becomes one "name=value" observation token (mirroring
// the C# JsonElement.ToString() extraction). Both models are in-memory,
// deterministic, and thread-safe.
//
// The C# dictionaries use StringComparer.OrdinalIgnoreCase, which matches
// case-insensitively but RETURNS the originally-stored key. We reproduce that
// with CiCounter: keyed by lower-case for lookup, remembering the first-seen
// original spelling so the predicted outcome keeps its original casing.

package com.bhengubv.circleai.companion.reasoning

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.max

// ---------------------------------------------------------------------------
// CausalPrediction
// ---------------------------------------------------------------------------

/**
 * A predicted [outcome] with its [probability] and the [supportingFactors] that
 * drove it. Mirrors the C# `CausalPrediction` record.
 */
data class CausalPrediction(
    val outcome: String,
    val probability: Double,
    val supportingFactors: List<String>,
)

// ---------------------------------------------------------------------------
// IWorldModel
// ---------------------------------------------------------------------------

/** World model + causal reasoning: predict an outcome for a scenario. */
interface IWorldModel {
    /**
     * Predict the most probable outcome for [scenarioJson] (a JSON object whose
     * properties are the observed variables).
     */
    suspend fun predictAsync(scenarioJson: String): CausalPrediction
}

// ---------------------------------------------------------------------------
// Case-insensitive counter (mirrors StringComparer.OrdinalIgnoreCase)
// ---------------------------------------------------------------------------

/**
 * A thread-safe count-per-key map that matches keys case-insensitively but
 * remembers the first-seen original spelling of each key — exactly how a .NET
 * `Dictionary<string,long>` created with `StringComparer.OrdinalIgnoreCase`
 * behaves. [increment] adds to a key; [snapshot] exposes (originalKey, count).
 */
internal class CiCounter {
    private data class Cell(val original: String, var count: Long)

    private val cells = ConcurrentHashMap<String, Cell>()

    fun increment(key: String, by: Long = 1L) {
        cells.compute(key.lowercase()) { _, existing ->
            if (existing == null) Cell(key, by) else existing.also { it.count += by }
        }
    }

    fun getCount(key: String): Long = cells[key.lowercase()]?.count ?: 0L

    fun total(): Long = cells.values.sumOf { it.count }

    fun size(): Int = cells.size

    val isEmpty: Boolean get() = cells.isEmpty()

    /** (originalKey, count) for every distinct key. */
    fun snapshot(): List<Pair<String, Long>> = cells.values.map { it.original to it.count }
}

// ---------------------------------------------------------------------------
// Shared observation extraction
// ---------------------------------------------------------------------------

private val EXTRACT_JSON: Json = Json { isLenient = true }

/**
 * Turns a scenario JSON object into a flat list of `name=value` observation
 * tokens, byte-for-byte matching the C# `ExtractObservations`
 * (`prop.Name + "=" + prop.Value.ToString()`). The rendering reproduces .NET's
 * `JsonElement.ToString()` exactly, which differs from a naive `toString()`:
 *   - string  → the unquoted content        ("rain")
 *   - number  → the raw JSON number text     ("5", "3.5")
 *   - `true`  → `True`   and `false` → `False`   (capitalised, .NET bool.ToString)
 *   - `null`  → the empty string             ("")
 *   - object/array → minified JSON           ({"a":1}, [1,2])
 *
 * Returns an empty list when [scenarioJson] is blank, not a JSON object, or not
 * parseable — the C# code catches `JsonException` and returns empty.
 */
internal fun extractObservations(scenarioJson: String?): List<String> {
    if (scenarioJson.isNullOrBlank()) return emptyList()
    return try {
        val root = EXTRACT_JSON.parseToJsonElement(scenarioJson)
        if (root !is JsonObject) return emptyList()
        root.entries.map { (name, value) -> "$name=${renderJsonElement(value)}" }
    } catch (_: Exception) {
        emptyList()
    }
}

/** Reproduces .NET `JsonElement.ToString()` for a kotlinx [value]. */
private fun renderJsonElement(value: kotlinx.serialization.json.JsonElement): String = when (value) {
    is JsonPrimitive -> when {
        value is kotlinx.serialization.json.JsonNull -> ""       // .NET: null -> ""
        value.isString -> value.content                          // unquoted string
        value.content == "true" -> "True"                        // .NET: bool.ToString()
        value.content == "false" -> "False"
        else -> value.content                                    // number: raw text
    }
    else -> value.toString()                                     // object/array: minified JSON
}

// ---------------------------------------------------------------------------
// FrequencyWorldModel
// ---------------------------------------------------------------------------

/**
 * Learns P(outcome | observation) as raw frequencies. At predict time it tallies
 * every outcome ever seen alongside any of the scenario's observations and
 * returns the most-frequent one, with probability = topCount / totalTally.
 *
 * Observation and outcome keys are matched case-insensitively (C#
 * `OrdinalIgnoreCase`). When no evidence matches, returns `("unknown", 0.5, …)`.
 */
class FrequencyWorldModel : IWorldModel {
    // observation -> { outcome -> count } (both case-insensitive).
    private val counts = ConcurrentHashMap<String, CiCounter>()

    /** Tell the model: when these observations happen, this outcome was seen. */
    fun observe(observations: Iterable<String>, outcome: String) {
        require(outcome.isNotBlank()) { "outcome required" }
        for (obs in observations) {
            val inner = counts.getOrPut(obs.lowercase()) { CiCounter() }
            inner.increment(outcome)
        }
    }

    override suspend fun predictAsync(scenarioJson: String): CausalPrediction {
        val observations = extractObservations(scenarioJson)
        // tally keyed case-insensitively, preserving the original outcome spelling.
        val tally = CiCounter()
        val supporters = ArrayList<String>()
        for (obs in observations) {
            val inner = counts[obs.lowercase()] ?: continue
            supporters.add(obs)
            for ((outcome, n) in inner.snapshot()) {
                tally.increment(outcome, n)
            }
        }
        if (tally.isEmpty) return CausalPrediction("unknown", 0.5, supporters)
        val total = tally.total()
        val top = tally.snapshot().maxByOrNull { it.second }!!
        return CausalPrediction(top.first, top.second.toDouble() / total, supporters)
    }
}

// ---------------------------------------------------------------------------
// BayesianWorldModel
// ---------------------------------------------------------------------------

/**
 * A small online-learning Naive Bayes classifier over (observations → outcome)
 * pairs. At predict time, for every seen outcome it evaluates
 *   log P(outcome | obs) = log P(outcome) + Σ log P(obs_i | outcome)
 * with Laplace smoothing (strength [laplaceAlpha]), then softmaxes the
 * log-posteriors to a normalised probability for the top outcome.
 *
 * When there are no observations or no evidence yet, returns
 * `("unknown", 0.5, [])`.
 */
class BayesianWorldModel(private val laplaceAlpha: Double = 1.0) : IWorldModel {

    init {
        require(laplaceAlpha > 0) { "laplaceAlpha must be > 0" }
    }

    private val outcomeCounts = CiCounter()
    // outcome (lower-cased key) -> { observation -> count } (case-insensitive).
    private val condCounts = ConcurrentHashMap<String, CiCounter>()
    private val vocab = HashSet<String>()
    private val vocabLock = Any()

    @Volatile
    private var totalObservations: Long = 0
    private val totalLock = Any()

    /** Update the model with one (observations → outcome) example. */
    fun observe(observations: Iterable<String>, outcome: String) {
        require(outcome.isNotBlank()) { "outcome required" }

        outcomeCounts.increment(outcome)
        synchronized(totalLock) { totalObservations++ }

        val cond = condCounts.getOrPut(outcome.lowercase()) { CiCounter() }
        for (obs in observations) {
            if (obs.isBlank()) continue
            cond.increment(obs)
            synchronized(vocabLock) { vocab.add(obs.lowercase()) }
        }
    }

    override suspend fun predictAsync(scenarioJson: String): CausalPrediction {
        val observations = extractObservations(scenarioJson)
        if (observations.isEmpty() || outcomeCounts.isEmpty) {
            return CausalPrediction("unknown", 0.5, emptyList())
        }

        val vocabSize = max(1, synchronized(vocabLock) { vocab.size })
        val totalEx = max(1L, totalObservations)
        val numOutcomes = outcomeCounts.size()

        val scored = ArrayList<Pair<String, Double>>(numOutcomes)
        for ((outcome, outcomeCount) in outcomeCounts.snapshot()) {
            // Log P(outcome) — Laplace-smoothed prior.
            val logPrior = ln((outcomeCount + laplaceAlpha) / (totalEx + laplaceAlpha * numOutcomes))

            val cond = condCounts[outcome.lowercase()]
            val totalForOutcome = cond?.total() ?: 0L
            var logLikelihood = 0.0
            for (obs in observations) {
                val n = cond?.getCount(obs) ?: 0L
                val p = (n + laplaceAlpha) / (totalForOutcome + laplaceAlpha * vocabSize)
                logLikelihood += ln(p)
            }
            scored.add(outcome to (logPrior + logLikelihood))
        }

        // Softmax over log-posteriors for a normalised probability.
        val maxLogPost = scored.maxOf { it.second }
        val expSum = scored.sumOf { exp(it.second - maxLogPost) }
        val top = scored.maxByOrNull { it.second }!!
        val prob = exp(top.second - maxLogPost) / expSum
        return CausalPrediction(top.first, prob, observations)
    }
}

File: python/src/circle_ai/companion/world_model.py
# companion/world_model.py
#
# IWorldModel implementations. Ported from CircleAI.Companion — the C# reference:
#
#   * BayesianWorldModel  (BayesianWorldModel.cs)          — online Naive Bayes
#   * FrequencyWorldModel (HerJarvisRealImplementations.cs) — frequency tally
#
# BayesianWorldModel is a real (small but honest) probabilistic graphical model:
# an online-learning Naive Bayes classifier over (observations -> outcome) pairs.
# At predict time, for every previously-seen outcome:
#     P(outcome | obs) ~ P(outcome) * prod_i P(obs_i | outcome)
# using Laplace smoothing so unseen pairs don't zero out, then softmax over the
# log-posteriors for a normalised probability.
#
# FrequencyWorldModel is the simpler predecessor: it learns P(outcome|observation)
# as a raw frequency tally.
#
# Both extract observations from a scenario JSON object as ``name=value`` pairs.
# The ``value`` rendering matches System.Text.Json's ``JsonElement.ToString()``
# byte-for-byte (booleans render as ``True``/``False``, null as empty string,
# nested objects/arrays as compact JSON) — see ``_element_to_string``.

from __future__ import annotations

import json
import math
import threading
from typing import Iterable, List, Optional, Tuple

from .herjarvis_contracts import CausalPrediction, IWorldModel


# ── shared JSON observation extraction ────────────────────────────────────


def _element_to_string(value: object) -> str:
    """Render a parsed-JSON value the way System.Text.Json's
    ``JsonElement.ToString()`` does.

    * string  -> the unescaped value, no quotes
    * bool    -> ``True`` / ``False``  (.NET capitalisation — Python's ``str``)
    * None    -> empty string
    * number  -> its compact numeric text
    * object/array -> compact JSON (``{"x":1}`` / ``[1,2]``, no spaces)
    """
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    if isinstance(value, bool):
        # Python's str(True) == "True", matching .NET's Boolean.ToString().
        return str(value)
    if isinstance(value, (int, float)):
        return json.dumps(value)
    # object / array -> compact JSON with no whitespace (matches STJ output).
    return json.dumps(value, separators=(",", ":"), ensure_ascii=False)


def _extract_observations(scenario_json: str) -> List[str]:
    """Parse a scenario JSON *object* into ``name=value`` observation strings.

    Returns an empty list when the JSON is not an object or fails to parse
    (JSON-syntax errors). A ``None`` input raises ``ValueError`` — this mirrors
    ``FrequencyWorldModel``, whose C# ``JsonDocument.Parse(null)`` throws
    ``ArgumentNullException`` (it catches only ``JsonException``). Callers that
    want the null-safe behaviour (``BayesianWorldModel``) guard for ``None``
    before calling.
    """
    if scenario_json is None:
        # Matches C# ArgumentNullException from JsonDocument.Parse(null).
        raise ValueError("scenario_json required")
    try:
        root = json.loads(scenario_json)
    except ValueError:  # JSON syntax error == C# JsonException -> empty
        return []
    if not isinstance(root, dict):
        return []
    return [f"{name}={_element_to_string(val)}" for name, val in root.items()]


# ── case-insensitive counting maps (StringComparer.OrdinalIgnoreCase) ──────


class _CiCounter:
    """A case-insensitive string->count map.

    Keys compare case-insensitively (like ``StringComparer.OrdinalIgnoreCase``);
    the first-seen original casing of each key is preserved for iteration, which
    is what a C# ``ConcurrentDictionary``/``Dictionary`` keyed by that comparer
    exposes via its ``Key``.
    """

    __slots__ = ("_map",)

    def __init__(self) -> None:
        # lower-key -> (original-key, count)
        self._map: dict[str, Tuple[str, int]] = {}

    def add(self, key: str, amount: int = 1) -> None:
        lk = key.lower()
        existing = self._map.get(lk)
        if existing is None:
            self._map[lk] = (key, amount)
        else:
            self._map[lk] = (existing[0], existing[1] + amount)

    def get(self, key: str) -> int:
        existing = self._map.get(key.lower())
        return existing[1] if existing is not None else 0

    def total(self) -> int:
        return sum(v[1] for v in self._map.values())

    def is_empty(self) -> bool:
        return len(self._map) == 0

    def __len__(self) -> int:
        return len(self._map)

    def items(self) -> Iterable[Tuple[str, int]]:
        """(original-key, count) pairs."""
        return [(orig, cnt) for (orig, cnt) in self._map.values()]


# ======================================================================
# BayesianWorldModel — online-learning Naive Bayes over (obs -> outcome).
# ======================================================================
class BayesianWorldModel(IWorldModel):
    """Online Naive Bayes world model with Laplace smoothing.

    Mirrors ``CircleAI.Companion.BayesianWorldModel``.
    """

    __slots__ = (
        "_outcome_counts",
        "_cond_counts",
        "_vocab",
        "_total_observations",
        "_alpha",
        "_lock",
    )

    def __init__(self, laplace_alpha: float = 1.0) -> None:
        if laplace_alpha <= 0:
            raise ValueError("laplace_alpha must be > 0")
        self._outcome_counts = _CiCounter()
        # outcome -> (observation -> count)
        self._cond_counts: dict[str, _CiCounter] = {}
        self._vocab: set[str] = set()  # case-insensitive vocab (lowercased)
        self._total_observations: int = 0
        self._alpha: float = laplace_alpha
        self._lock = threading.Lock()

    def observe(self, observations: Iterable[str], outcome: str) -> None:
        """Update the model with one (observations -> outcome) example."""
        if observations is None:  # type: ignore[comparison-overlap]
            raise ValueError("observations required")
        if outcome is None or len(outcome.strip()) == 0:
            raise ValueError("outcome required")

        with self._lock:
            self._outcome_counts.add(outcome, 1)
            self._total_observations += 1

            key = outcome.lower()
            cond = self._cond_counts.get(key)
            if cond is None:
                cond = _CiCounter()
                self._cond_counts[key] = cond
            for obs in observations:
                if obs is None or len(obs.strip()) == 0:
                    continue
                cond.add(obs, 1)
                self._vocab.add(obs.lower())

    async def predict_async(
        self, scenario_json: str, *, ct: Optional[object] = None
    ) -> CausalPrediction:
        # BayesianWorldModel.ExtractObservations guards IsNullOrWhiteSpace and
        # returns empty (never throws) — so null/blank -> no observations here.
        if scenario_json is None or len(scenario_json.strip()) == 0:
            observations: List[str] = []
        else:
            observations = _extract_observations(scenario_json)
        with self._lock:
            if len(observations) == 0 or self._outcome_counts.is_empty():
                return CausalPrediction("unknown", 0.5, ())

            vocab_size = max(1, len(self._vocab))
            total_ex = max(1, self._total_observations)
            outcome_count_n = len(self._outcome_counts)

            scored: List[Tuple[str, float]] = []
            for outcome, outcome_count in self._outcome_counts.items():
                # Log P(outcome) — Laplace-smoothed prior.
                log_prior = math.log(
                    (outcome_count + self._alpha)
                    / (total_ex + self._alpha * outcome_count_n)
                )

                cond = self._cond_counts.get(outcome.lower())
                total_for_outcome = cond.total() if cond is not None else 0
                log_likelihood = 0.0
                for obs in observations:
                    n = cond.get(obs) if cond is not None else 0
                    p = (n + self._alpha) / (
                        total_for_outcome + self._alpha * vocab_size
                    )
                    log_likelihood += math.log(p)
                scored.append((outcome, log_prior + log_likelihood))

            # Softmax over log-posteriors for normalised probability.
            max_log_post = max(s[1] for s in scored)
            exp_sum = sum(math.exp(s[1] - max_log_post) for s in scored)
            top = max(scored, key=lambda s: s[1])
            prob = math.exp(top[1] - max_log_post) / exp_sum
            return CausalPrediction(top[0], prob, tuple(observations))


# ======================================================================
# FrequencyWorldModel — learn P(outcome|observation) from evidence.
# ======================================================================
class FrequencyWorldModel(IWorldModel):
    """Frequency-tally world model.

    Mirrors ``CircleAI.Companion.HerJarvis.FrequencyWorldModel``.
    """

    __slots__ = ("_counts", "_lock")

    def __init__(self) -> None:
        # observation -> (outcome -> count), both case-insensitive
        self._counts: dict[str, _CiCounter] = {}
        self._lock = threading.Lock()

    def observe(self, observations: Iterable[str], outcome: str) -> None:
        """Tell the model: when these observations happen, this outcome was seen."""
        if observations is None:  # type: ignore[comparison-overlap]
            raise ValueError("observations required")
        if outcome is None or len(outcome.strip()) == 0:
            raise ValueError("outcome required")
        with self._lock:
            for obs in observations:
                lk = obs.lower()
                inner = self._counts.get(lk)
                if inner is None:
                    inner = _CiCounter()
                    self._counts[lk] = inner
                inner.add(outcome, 1)

    async def predict_async(
        self, scenario_json: str, *, ct: Optional[object] = None
    ) -> CausalPrediction:
        # NOTE: FrequencyWorldModel does NOT null-guard the JSON before parsing
        # (unlike BayesianWorldModel); a None/blank input simply yields no
        # observations here, exactly as the C# try/catch does.
        observations = _extract_observations(scenario_json)
        with self._lock:
            tally = _CiCounter()
            supporters: List[str] = []
            for obs in observations:
                inner = self._counts.get(obs.lower())
                if inner is None:
                    continue
                supporters.append(obs)
                for outcome, cnt in inner.items():
                    tally.add(outcome, cnt)
            if tally.is_empty():
                return CausalPrediction("unknown", 0.5, tuple(supporters))
            total = tally.total()
            top = max(tally.items(), key=lambda kv: kv[1])
            return CausalPrediction(top[0], top[1] / total, tuple(supporters))


__all__ = [
    "BayesianWorldModel",
    "FrequencyWorldModel",
]
